File: audiocloud-api/src/media.rs
//! Communication with the on-site media library

use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::newtypes::{AppMediaObjectId, AppSessionId};
use crate::session::{SessionTrackChannels, SessionTrackMediaFormat};
use crate::time::{now, Timestamp, Timestamped};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct MediaJobState {
    pub progress:    f64,
    pub retry:       usize,
    pub error:       Option<String>,
    pub in_progress: bool,
    pub updated_at:  Timestamp,
}

impl Default for MediaJobState {
    fn default() -> Self {
        Self { progress:    0.0,
               retry:       0,
               error:       None,
               in_progress: false,
               updated_at:  now(), }
    }
}

impl MediaJobState {
    pub fn is_finished_ok(&self) -> bool {
        !self.in_progress && self.error.is_none()
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct DownloadMedia {
    pub get_url:    String,
    pub notify_url: String,
    pub context:    String,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct MediaMetadata {
    pub channels:    SessionTrackChannels,
    pub format:      SessionTrackMediaFormat,
    pub seconds:     f64,
    pub sample_rate: usize,
    pub bytes:       u64,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct UploadToDomain {
    pub channels:    SessionTrackChannels,
    pub format:      SessionTrackMediaFormat,
    pub seconds:     f64,
    pub sample_rate: usize,
    pub bytes:       u64,
    pub url:         String,
    pub notify_url:  Option<String>,
    // typescript: any
    pub context:     Option<Value>,
}

impl UploadToDomain {
    pub fn metadata(&self) -> MediaMetadata {
        MediaMetadata { channels:    self.channels,
                        format:      self.format,
                        seconds:     self.seconds,
                        sample_rate: self.sample_rate,
                        bytes:       self.bytes, }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct DownloadFromDomain {
    pub url:        String,
    pub notify_url: Option<String>,
    // typescript: any
    pub context:    Option<Value>,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct ImportToDomain {
    pub path:        String,
    pub channels:    SessionTrackChannels,
    pub format:      SessionTrackMediaFormat,
    pub seconds:     f64,
    pub sample_rate: usize,
    pub bytes:       u64,
}

impl ImportToDomain {
    pub fn metadata(&self) -> MediaMetadata {
        MediaMetadata { channels:    self.channels,
                        format:      self.format,
                        seconds:     self.seconds,
                        sample_rate: self.sample_rate,
                        bytes:       self.bytes, }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct MediaDownload {
    pub download: DownloadFromDomain,
    pub state:    MediaJobState,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct MediaUpload {
    pub upload: UploadToDomain,
    pub state:  MediaJobState,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct MediaObject {
    pub id:       AppMediaObjectId,
    pub metadata: Option<MediaMetadata>,
    pub path:     Option<String>,
    pub download: Option<MediaDownload>,
    pub upload:   Option<MediaUpload>,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct UpdateMediaSession {
    pub media_objects: HashSet<AppMediaObjectId>,
    pub ends_at:       Timestamp,
}

#[derive(Serialize, Deserialize, Clone, Debug)]
pub enum MediaServiceCommand {
    SetSessionMedia {
        session_id: AppSessionId,
        media:      HashSet<AppMediaObjectId>,
    },
    DeleteSession {
        session_id: AppSessionId,
    },
}

File: js-api/src/media.ts
import { Static, Type } from "@sinclair/typebox";
import { AppMediaObjectId, AppSessionId } from "./new_types";
import { SessionTrackChannels, SessionTrackMediaFormat } from "./session";
import { JsonTimeStamp } from "./time";
import Option from "./utils/option";

export const MediaJobState = Type.Object({
    progress:       Type.Number(),
    retry:          Type.Integer(),
    error:          Option(Type.String()),
    in_progress:    Type.Boolean(),
    updated_at:     JsonTimeStamp
})
export type MediaJobState = Static<typeof MediaJobState>

export const DownloadMedia = Type.Object({
    get_url:        Type.String(),
    notify_url:     Type.String(),
    context:        Type.String()
})
export type DownloadMedia = Static<typeof DownloadMedia>

export const MediaMetadata = Type.Object({
    channels:       SessionTrackChannels,
    format:         SessionTrackMediaFormat,
    seconds:        Type.Number(),
    sample_rate:    Type.Integer(),
    bytes:          Type.Integer()
})
export type MediaMetadata = Static<typeof MediaMetadata>

export const ImportToDomain = Type.Object({
    path:           Type.String(),
    channels:       SessionTrackChannels,
    format:         SessionTrackMediaFormat,
    seconds:        Type.Number(),
    sample_rate:    Type.Integer(),
    bytes:          Type.Integer()
})
export type ImportToDomain = Static<typeof ImportToDomain>

export const UploadToDomain = Type.Object({
    channels:       SessionTrackChannels,
    format:         SessionTrackMediaFormat,
    seconds:        Type.Number(),
    sample_rate:    Type.Integer(),
    bytes:          Type.Integer(),
    url:            Type.String(),
    notify_url:     Option(Type.String()),
    context:        Option(Type.Any())
})
export type UploadToDomain = Static<typeof UploadToDomain>

export const DownloadFromDomain = Type.Object({
    url:            Type.String(),
    notify_url:     Option(Type.String()),
    context:        Option(Type.Any())
})
export type DownloadFromDomain = Static<typeof UploadToDomain>

export const MediaDownload = Type.Object({
    download:       DownloadFromDomain,
    state:          MediaJobState
})
export type MediaDownload = Static<typeof MediaDownload>

export const MediaUpload = Type.Object({
    download:       UploadToDomain,
    state:          MediaJobState
})
export type MediaUpload = Static<typeof MediaUpload>

export const MediaObject = Type.Object({
    id:             AppMediaObjectId,
    metadata:       Option(MediaMetadata),
    path:           Option(Type.String()),
    download:       Option(MediaDownload),
    upload:         Option(MediaUpload)
})
export type MediaObject = Static<typeof MediaObject>

export const UpdateMediaSession = Type.Object({
    media_objects:  Type.Array(AppMediaObjectId),
    ends_at:        JsonTimeStamp
})

export const MediaServiceCommand = Type.Union([
    Type.Object({
        "set_session_media":    AppSessionId,
        "media":                Type.Array(AppMediaObjectId)
    }),
    Type.Object({
        "delete_session":       AppSessionId
    })
])
export type MediaServiceCommand = Static<typeof MediaServiceCommand>
